Catching up | [Main] Debug mode, parse utils, MIME | [Bridges] Add/Improve 20 bridges

ORelio opened the pull request:
Hello everyone,

It's been a while since my last contribution! Due to new requirements/dependencies I could not install RSS-Bridge updates anymore on my old server and had to maintain my old RSS-Bridge instance detached from the main code base. This is now resolved and I'm trying to catch up with the latest changes made to RSS-Bridge.

This pull request aims to share improvements for the main code base as well as for some bridges. I did my best to follow the new contridution guidelines and tried to accomodate with the new coding style rules, but in order to avoid opening 23 pull requests at once, I separated each contribution - e.g. 1 bridge or 1 feature - in individual commits inside this pull request. Apologies for the review workload :confused:

Here is a summary of the changes:

## Main

* **[Debug mode]** fix PHP errors not showing, fix PHP notices, improve warning message in bridge list
* **[Parse utilities]** Instead of copying the sames functions everywhere, move them to html.php
* **[MIME type]** Auto-handle MIME type for enclosures based on file extension

## Bridges

* **[New]** Anidex
* **[Improved]** Anime-Ultime
* **[Repaired]** CNET
* **[Improved]** Dilbert
* **[Repaired]** EstCeQuonMetEnProd
* **[Improved]** Facebook
* **[Improved]** Futura-Sciences
* **[Repaired]** GBATemp
* **[Repaired]** JapanExpo
* **[Repaired]** LeMondeInformatique
* **[Repaired]** Nextgov
* **[Repaired]** NextInpact
* **[New]** NyaaTorrents
* **[Improved]** Releases3DS
* **[Repaired]** TheHackerNews
* **[Repaired]** WeLiveSecurity
* **[Improved]** WordPress
* **[Improved]** YGGTorrent
* **[Repaired]** ZDNet
* **[Improved]** FeedExpander

For each change, please see description attached to the associated commit. Should you have any question or require any change before merge, please let me know. :slightly_smiling_face:

ORelio committed:
Debug mode improvements

 - Improve debug warning message
 - Restore error reporting in debug mode
 - Fix 'notice' messages for unset fields

ORelio committed:
Add parsing utility functions

html.php
 - extractFromDelimiters
 - stripWithDelimiters
 - stripRecursiveHTMLSection
 - markdownToHtml (partial)

bridges
 - remove now-duplicate functions
 - call functions from html.php instead

ORelio committed:
[Anidex] New bridge

Anime torrent tracker

ORelio committed:
[Anime-Ultime] Restore thumbnail

ORelio committed:
[CNET] Recreate bridge

Full rewrite as the previous one was broken

ORelio committed:
[Dilbert] Minor URI fix

Use new self::URI property

ORelio committed:
[EstCeQuonMetEnProd] Fix content extraction

Bridge was broken

ORelio committed:
[Facebook] Fix "SpSonsSoriSsés" label

... which was taking space in item title

ORelio committed:
[Futura-Sciences] Use HTTPS, More cleanup

Use HTTPS as FS now offer HTTPS
Clean additional useless HTML elements

ORelio committed:
[GBATemp] Multiple fixes

- Fix categories: missing "break" statements
- Restore thumbnail as enclosure
- Fix date extraction
- Fix user blog post extraction
- Use getSimpleHTMLDOMCached

ORelio committed:
[JapanExpo] Fix bridge, HTTPS, thumbnails

- Fix getSimpleHTMLDOMCached call
- Upgrade to HTTPS as JE now offers HTTPS
- Restore thumbnails as enclosures

ORelio committed:
[LeMondeInformatique] Fix bridge, HTTPS

- Upgrade to HTTPS as LMI now offers HTTPS
- Restore thumbnails using small images
- Fix content extraction
- Fix text encoding issue

ORelio committed:
[Nextgov] Fix content extraction

- Restore thumbnail and use small image
- Field extraction fixes

ORelio committed:
[NextInpact] Add categories and filtering by type

- Offer all RSS feeds
- Allow filtering by article type
- Implement extraction for brief articles
- Remove article limit, many brief articles are publied all at once

ORelio committed:
[NyaaTorrents] New bridge

Anime torrent tracker

ORelio committed:
[Releases3DS] Cache content, restore thumbnail

- Use getSimpleHTMLDOMCached
- Restore thumbnail as enclosure

ORelio committed:
[TheHackerNews] Fix bridge

 - Fix content extraction including article body
 - Restore thumbnail as enclosure

ORelio committed:
[WeLiveSecurity] HTTPS, Fix content extraction

- Upgrade to HTTPS as WLS now offers HTTPS
- Fix content extraction including article body

ORelio committed:
[WordPress] Reduce timeout, more content selectors

- Reduce timeout to use default one (1h)
- Add new content selector (articleBody)
- Find thumbnail and set as enclosure
- Fix <script> cleanup

ORelio committed:
[YGGTorrent] Increase limit, use cache

- Increase item limit as uploads are very frequent
- Use getSimpleHTMLDOMCached

ORelio committed:
[ZDNet] Rewrite with FeedExpander

- Upgrade to HTTPS as ZD now offers HTTPS
- Use FeedExpander for secondary fields
- Fix content extraction for article body

ORelio committed:
[Main] Handle MIME type for enclosures

Many feed readers will ignore enclosures (e.g. thumbnails) with no MIME type. This commit adds automatic MIME type detection based on file extension (which may be inaccurate but is the only way without fetching the content).

One can force enclosure type using #.ext anchor (hacky, needs improving)

ORelio committed:
[FeedExpander] Improve field extraction

- Add support for passing enclosures
- Improve author and uri extraction
- Fix 'notice' PHP error messages

ORelio committed:
[Pull] Coding style fixes for #802

teromene commented on bridges/AnidexBridge.php:
```diff
@@ -0,0 +1,205 @@
+				$item_uri = self::URI . 'torrent/'.$torrent_id;
+
+				//Retrieve full description from torrent page
+				if ($item_html = getSimpleHTMLDOMCached($item_uri)) {
```
You should fail if the condition is false, else it might enter an infinite loop (or at least move the $limit++ up)

ORelio replied:
This bridge works by retrieving all links in the page and processing only links containing a specific pattern, in a limit of 20. Some links *will* not pass the conditions and are ignored on purpose. I think this behavior is not likely to cause an infinite loop since the amount of links in the page would need to be infinite too and Simple DOM Parser would fail before that. How about moving `$limit++` right before calling `getSimpleHTMLDOMCached`?

logmanoriginal replied:
See my comment below. You can simply check for the limit after adding items to `$this->items`:

```PHP
if(count($this->items) >= 10) {
    break;
}
```

That way, either the limit is reached and the loop ends, or the loop continues until all items were checked and terminates normally.

ORelio replied:
Thanks, changing to that. :+1:

teromene commented on bridges/AnimeUltimeBridge.php:
```diff
@@ -91,6 +97,8 @@ public function collectData(){
 								strpos($item_description, '<div id="table">')
 							);
+
+							// Convert relative image src into absolute image src, remove line breaks
```
We have an internal function for this I think

ORelio replied:
Yes, in `html.php` (`defaultLinkTo`). In the case of this bridge, I only added a comment and did not touch existing code, which could have been replaced by a call to this function. However, `defaultLinkTo` does not seems ready yet, as it will not work for all websites because some sites are using a mix of:

 - Full links `https://site.com/some/resource.html`
 - Links relative to the root `/some/resource.html`
 - Links relative to the resource being viewed eg `otherpage.html`

Here is an [example](https://www.futura-sciences.com/tech/actualites/tablette-magicscroll-curieuse-tablette-smartphone-inspiree-parchemin-72660/) of a page containing both links relative to the root and absolute links. The function would prepend domain names to links already containing it, thus breaking some valid links.

In order to solve this problem, the function would need both the website URI e.g. `https://site.com` *and* the current resource URI e.g. `/some/resource.html` (or full URI, that would work too) in order to properly rebuild links relative to the resource being viewed.

If you are interested in that, I could try rewriting the function :slightly_smiling_face:

logmanoriginal replied:
`defaultLinkTo` should work if provided the link of the resource being viewed. `php-urljoin`, which is used internally, is supposed to handle full, relative from provided link and relative from root situations.

If it's broken, please add an example to the test cases: https://github.com/fluffy-critter/php-urljoin/blob/master/tests/tests.php

ORelio replied:
OK, I'm wrong, did not notice that `php-urljoin` actually does that. Fixing the bridge and adding documentation above `defaultLinkTo` to explain what it does :+1:

logmanoriginal replied:
Good point, thanks for taking care of that :+1: 

Although some functions are documented, the core is missing a lot of documentation. Maybe we can get this up to standard (i.e. for phpDocumentor), but that is a separate task for the future. Most functions are explained in the [Wiki](https://github.com/RSS-Bridge/rss-bridge/wiki) at least.

teromene commented on bridges/CNETBridge.php:
```diff
@@ -0,0 +1,109 @@
+					$item['enclosures'] = array($article_thumbnail);
+					$item['content'] = $article_content;
+					$this->items[] = $item;
+					$limit++;
```
Same as before, the limit should be just after line 62

ORelio replied:
In the case of this bridge, yes it's possible since 1 checked item = 1 processed item. Same as before, I think there is no `continue` or similar statement that could skip increasing the limit, but it's safer to increase it as soon as possible, indeed. I'll fix that.

logmanoriginal replied:
You can get rid of the if condition by simply checking the array size and breaking the loop:

```PHP
if(count($this->items) >= 10) {
    break;
}
```

ORelio replied:
Thanks, changing to that. :+1:

teromene commented on bridges/GBAtempBridge.php:
```diff
@@ -76,70 +84,56 @@ public function collectData(){
-						'"'
-					)
-				);
+				$img = $this->getURI().$newsItem->find('img', 0)->src . '#.image';
```
```PHP
 				$img = $this->getURI() . $newsItem->find('img', 0)->src . '#.image'; 
```

ORelio replied:
Good catch :+1:

teromene commented on bridges/NyaaTorrentsBridge.php:
```diff
@@ -0,0 +1,129 @@
+
+		// Build Search URL from user-provided parameters
+		$search_url = self::URI . '?s=id&o=desc';
+		foreach (array('f', 'c') as $param_name) {
```
You probably can use ```http_build_query```

ORelio replied:
Completely forgot PHP had this function. Thanks. Remark: In that case, the bridge won't filter or omit unset parameters. That should not be an issue since now RSS-Bridge checks `list` parameters are set within a range of allowed values.

teromene commented on bridges/NyaaTorrentsBridge.php:
```diff
@@ -0,0 +1,129 @@
+			if (strpos($link->href, '/view/') === 0 && !in_array($link->href, $results))
+				$results[] = $link->href;
+		if (empty($results))
+			returnServerError('No results from Nyaa: '.$url, 500);
```
```PHP
returnServerError('No results from Nyaa: ' . $url, 500);
```

ORelio replied:
Will fix :+1:

teromene commented on bridges/NyaaTorrentsBridge.php:
```diff
@@ -0,0 +1,129 @@
+
+			//Limit total amount of requests and ignore entries without valid torrent ID
+			if ($limit < 20 && $torrent_id != 0 && ctype_digit($torrent_id)) {
+
```
Same as before for the limit

ORelio replied:
This bridge works the same way as AnidexBridge (see discussion above). I can move `$limit++` right before `getSimpleHTMLDOMCached`.

ORelio replied:
Applying the same recommendation from @LogMANOriginal as `AnixedBridge`.
```php
if(count($this->items) >= 20) {
	break;
}
```

teromene commented on bridges/WeLiveSecurityBridge.php:
```diff
@@ -3,37 +3,24 @@ class WeLiveSecurityBridge extends FeedExpander {
 		$article_html = getSimpleHTMLDOMCached($item['uri']);
 		if(!$article_html) {
-			$item['content'] .= '<p>Could not request ' . $this->getName() . ': ' . $item['uri'] . '</p>';
+			$item['content'] .= '<p><em>Could not request ' . $this->getName() . ': ' . $item['uri'] . '</p>';
```
```returnServerError``` ?

ORelio replied:
When rewriting some bridges, I saw that this behavior was implemented, especially in bridges implementing `FeedExpander`. It seems a good idea as it helps to keep the bridge functional even if requests to page with full contents fails or parsing did not work due to changes on the website.

Let's imagine the following scenario:
 - Site X provides feed F, which is truncated
 - Bridge implements FeedExpander using feed F
 - Site X makes some changes, breaking the ability to request/parse pages
 - With failsave behavior, the bridge fallbacks to just passing fields from feed F
 - With `returnServerError`, the bridge just won't work after changes on site X

So this bridge uses the following approach:
 - Error on article list: `returnServerError`
 - Error on article content: try to keep going

logmanoriginal replied:
I think this is actually a good idea. `$item['content']` already contains the contents from the original feed (notice the `.=`), so there is no harm adding a notice and continue with what we have.

However, you should really close the `em` tag:

```PHP
$item['content'] .= '<p><em>Could not request ' . $this->getName() . ': ' . $item['uri'] . '</em></p>';
```

ORelio replied:
Good catch about the missing `</em>` :open_mouth: Fixing that.

teromene commented on index.php:
```diff
@@ -7,16 +7,23 @@
   can allow a malicious client to retrieve data about your server and hammer
   a provider throught your rss-bridge instance.
 */
+error_reporting(0);
```
We should let the default value as authority if we don't activate debug mode

ORelio replied:
Indeed and I agree with you. Actually this line is already present in index.php:

https://github.com/RSS-Bridge/rss-bridge/blob/master/index.php#L39

Obviously, this was breaking debug mode so all I did was moving this line **above** the debug mode checks in order to repair debug. I can simply remove that line if that's ok with you.

Edit: According to `git blame`, the line was introduced in 937ea492716b9251211a2a0f73f3afd1ac871505.

logmanoriginal replied:
> Actually this line is already present in index.php

Guilty as charged :smile: 

> I can simply remove that line if that's ok with you.

Please do so, there is no reason to keep it (hopefully :pray:)

ORelio replied:
After further research, actually, you just moved the line in 937ea492716b9251211a2a0f73f3afd1ac871505.
It was introduced way before as per 927b04dfef6c363674cfd8e727cd1c34d5485d68 :wink:

The only benefit from this line is that a user will never need PHP error messages... I guess. And someone wanting to debug will enable debug mode anyway.

Edit: Removing the line.

teromene commented on bridges/AnidexBridge.php:
```diff
@@ -0,0 +1,205 @@
+					$item_desc = $item_html->find('div.panel-body', 0);
+					$item_author = trim($item_html->find('span.fa-user', 0)->parent()->plaintext);
+					$item_date = strtotime(trim($item_html->find('span.fa-clock', 0)->parent()->plaintext));
+					$item_image = $this->getURI().'images/user_logos/default.png';
```
```PHP
 					$item_image = $this->getURI() . 'images/user_logos/default.png'; 
```

ORelio replied:
Will fix :+1:

teromene commented on bridges/NyaaTorrentsBridge.php:
```diff
@@ -0,0 +1,129 @@
+					$item_date = intval(extractFromDelimiters($item_html->outertext, 'data-timestamp="', '"'));
+
+					//Retrieve image for thumbnail or generic logo fallback
+					$item_image = $this->getURI().'static/img/avatar/default.png';
```
```PHP
 					$item_image = $this->getURI() . 'static/img/avatar/default.png'; 
```

ORelio replied:
Will fix :+1:

teromene commented on lib/html.php:
```diff
@@ -51,3 +51,95 @@ function defaultLinkTo($content, $server){
+		$string = preg_replace('/ (src|href)="([^"]+)<\/i>([^"]+)"/U', ' $1="$2_$3"', $string, -1, $count);
+	}
+
+	return '<div>'.trim($string).'</div>';
```
```PHP
 	return '<div>' . trim($string) . '</div>'; 
```

ORelio replied:
Will fix :+1:

teromene reviewed (changes requested):
Thank you for this huge PR !
Here are a few comments on the changes.
Moreover, I don't know whether having a static image or the article's image in the enclosures is the right way of using the system ?

ORelio commented:
Hi @teromene , thank you for the detailed review :slightly_smiling_face:

Regarding the enslosures thing, I think this needs improvements indeed (see explanations below). Regarding other issues, I'll address them directly by responding to your remarks.

The enclosure usage for thumbnails is a continuation of discussion in #340 . My feed reader (Feedly) has thumbnail image support, and I think other readers might do as well. The thumbnail helps having a more clear overview of the article when scrolling through a daily feed with hundreds of items, I find it very useful.

By default, Feedly (and maybe other readers) will use a random image from the article using some kind of algorithm. This often results in using a non-relevant image. However, if an enclosure of MIME type `image/<anything>` is present, that enclosure is used instead. For this reason, if the article has an image clearly set as the article image by the author (e.g. `wp-post-image` in WordPress), I'm setting it as an enclosure to explicitely tell the feed reader to use it instead of guessing what image to use. Some website publishers do as well in their feeds, by setting images as enclosures ([sample](https://www.futura-sciences.com/rss/actualites.xml)).

When no image is set or the website does not have that "article image" feature, however, I tried to provide one anyway - which is often not optimal - in order to avoid having an empty thumbnail in the feed reader. In that case, I'm using either the first image, or a fallback image (website logo or similar) to use as a placeholder in feed reader. Some RSS-Bridge users may not want this, so the following algorithm could be used instead:

````
if <some-setting-enabled> and <no-enclosures-in-item>:
        if <article-contains-images>:
            <do-nothing> // Let feed reader select an image by itself
            // Or maybe select first image for dumb readers?
        else:
            <provide a static placeholder image>
````

What do you think about this? I could add a setting (disabled by default) inside the new `config.php` file, then add a function e.g. in `content.php` or `html.php` which implements the algorithm for use in bridges:

````php
function placeholderThumbnail($item, $defaultImage)
````

The function would check the item, and attach an enclosure if the setting is enabled, item does not contain any image and no image enclosure is present, and would be called by bridges willing to take advantage of that feature / or this could be called automatically if a constant e.g. `LOGO` or `DEFAULT_IMAGE` is defined in the bridge (still suggesting).

Thanks in advance for any thought about this :slightly_smiling_face:

logmanoriginal commented on lib/BridgeList.php:
```diff
@@ -58,14 +58,21 @@ private static function getHeader() {
+	{$status}
 </header>
 EOD;
 	}
```
This is a very good idea! :+1: 

I suggest adding some more formatting to indicate the urgency of the warning. Here is slightly changed version based on your solution:

_Notice_: Below you'll find corresponding changes to the CSS

```PHP
	private static function getHeader() {
		$warning = '';

		if(defined('DEBUG') && DEBUG === true) {
			if(defined('DEBUG_INSECURE') && DEBUG_INSECURE === true) {
				$warning .= <<<EOD
<section class="critical-warning">Warning : Debug mode is active from any location, 
make sure only you can access RSS-Bridge.</section>
EOD;
			} else {
				$warning .= <<<EOD
<section class="warning">Warning : Debug mode is active from your IP address, 
your requests will bypass the cache.</section>
EOD;
			}
		}

		return <<<EOD
<header>
	<h1>RSS-Bridge</h1>
	<h2>Reconnecting the Web</h2>
	{$warning}
</header>
EOD;
	}
```

ORelio replied:
Agreed, this is even better. Changing. :+1:

logmanoriginal commented on static/style.css:
```diff
@@ -40,10 +40,10 @@ a:hover {
+	width: 40%;
+	background-color: #ffc600;
+	color: #5f5f5f;
 }
```
Based on my suggestion above, the warning message could indicate the urgency:

```CSS
 header > section.warning {
	width: 40%;
	background-color: #ffc600;
	color: #5f5f5f;
}

 header > section.critical-warning {
	 width: 40%;
	 background-color: #cf3e3e;
	 font-weight: bold;
	 color: white;
 }
```

![image](https://user-images.githubusercontent.com/5776685/45256517-ab4a3180-b397-11e8-9db3-75f2f2dd10c0.png)

***

![image](https://user-images.githubusercontent.com/5776685/45256526-bdc46b00-b397-11e8-804e-005283ce9d23.png)

ORelio replied:
Nice colors. Changing. :+1:

logmanoriginal commented on lib/html.php:
```diff
@@ -51,3 +51,95 @@ function defaultLinkTo($content, $server){
+		$string = preg_replace('/ (src|href)="([^"]+)<\/i>([^"]+)"/U', ' $1="$2_$3"', $string, -1, $count);
+	}
+
+	return '<div>'.trim($string).'</div>';
```
This is absolutely mind blowing and at the same time very hard to comprehend. Could you link a few sources or example data for each regex? I don't suppose there is a specification for this, right?

ORelio replied:
To make it clear, this function does not implement the full Markdown syntax, I only wrote this to make Markdown pretty much readable on `NyaaTorrentsBridge` without adding a new dependency to the project (which is preferable for proper implementation but maybe not for the sake of only one bridge).

According to Wikipedia, the initial specifications for Markdown are here: https://daringfireball.net/projects/markdown/syntax ... but the function I provide will only handle some of the most commonly used tags such as links, basic formatting and images. I did not leave comments for each line because I think the replacement string clearly states what each regex is for (`<img`, `<a`, `<b>` and so on).

Here is an explained example for each regex:

 - Images: https://regex101.com/r/JW9Evr/1
 - Links: https://regex101.com/r/eRGVe7/1
 - Bold: https://regex101.com/r/2p40Y0/1
 - Italic: https://regex101.com/r/xJkET9/1
 - Separator: https://regex101.com/r/ZBEqFP/1
 - Plain URL: https://regex101.com/r/2JHYwb/1
 - Site name: https://regex101.com/r/qIuKYE/1

Since some regex are also a subset of another regex, the most complex ones should be called first, e.g. images first, then links. However, I had an issue with links containing underscores causing italic text so I added a fixup regex (hacky):

 - Fixup regex `<i>`: https://regex101.com/r/NTRPf6/1
 - Fixup regex `</i>`: https://regex101.com/r/aNklRp/1

Hope this makes more clear what this function does. This was implemented as a simple but straightforward way of dealing with sites using javascript to render Markdown, but the only perfect way of handling it would be using another dependency for it. (Because implementing the full syntax requires more that a couple of regex :yum: )

logmanoriginal replied:
Thanks for the details. I agree that adding a dependency for Markdown support is not worth it right now. Having this function available, however, is very nice :+1: 

It would be great if you could add the links to the source code as well, so we know how it works in a year or two :smile: 

This function might become handy for other bridges as well (especially for GitHub and others that rely on Markdown), although I suppose some sites don't rely on JS that much.

ORelio replied:
OK, added that in a1dd50a89b6685161b68d4afaa4cdcda8527e437 :slightly_smiling_face:

logmanoriginal commented on lib/FeedExpander.php:
```diff
@@ -115,12 +115,27 @@ protected function load_ATOM_feed_data($content){
 		if(isset($feedItem->author)) $item['author'] = (string)$feedItem->author->name;
 		if(isset($feedItem->content)) $item['content'] = (string)$feedItem->content;
+
+		//When "link" field is present, URL is more relable than "id" field
```
*reliable

ORelio replied:
Woops. Fixing. :+1:

logmanoriginal commented on bridges/NextInpactBridge.php:
```diff
@@ -6,29 +6,105 @@ class NextInpactBridge extends FeedExpander {
-		$this->collectExpandableDatas(self::URI . 'rss/news.xml', 10);
+		$feed = $this->getInput('feed');
+		if (empty($feed))
+			$feed = 'news';
```
Instead of checking for an empty feed you can just specify the default value of the parameter itself and make it required:

```PHP
		'feed' => array(
			'name' => 'Feed',
			'type' => 'list',
			'required' => true,
			'values' => array(
				'Tous nos articles' => 'news',
				'Nos contenus en accès libre' => 'acces-libre',
				'Blog' => 'blog',
				'Bons plans' => 'bonsplans'
			),
			'defaultValue' => 'news'
		),
```

That way the function can be reduced to:

```PHP
	public function collectData(){
		$this->collectExpandableDatas(self::URI . 'rss/' . $this->getInput('feed') . '.xml');
	}
```

As bonus you receive an error message if the parameter is omitted:

![image](https://user-images.githubusercontent.com/5776685/45256913-8c4e9e00-b39d-11e8-9455-ce925053bc83.png)

ORelio replied:
The parameter is implemented this way for backward compatibility, because initially, this bridge did not have the `feed` parameter. If we make the field mandatory, a user upgrading RSS-Bridge will have to notice that and re-create a valid URL for the `NextInpactBridge`. Is it OK to keep that for compatibility?

logmanoriginal replied:
I see, that makes sense. With that in mind it makes sense not to use required parameters, so let's keep it how it is :+1:

logmanoriginal commented on bridges/AnidexBridge.php:
```diff
@@ -0,0 +1,205 @@
+	const MAINTAINER = 'ORelio';
+	const NAME = 'Anidex';
+	const URI = 'https://anidex.info/';
+	const CACHE_TIMEOUT = 3600; // 1h
```
This is the default value, so it can be omitted.

ORelio replied:
Indeed. Removing.

logmanoriginal commented on bridges/AnidexBridge.php:
```diff
@@ -0,0 +1,205 @@
+					$item['enclosures'] = array($item_image);
+					$item['content'] = $item_desc;
+					$this->items[] = $item;
+					$limit++;
```
Small improvement: You can get rid of $limit by checking the array size and breaking the loop:

```PHP
if(count($this->items) >= 20) {
    break;
}
```

ORelio replied:
Fixing :+1:

logmanoriginal reviewed (changes requested):
The amount of input in this PR is mind blowing, great job! :1st_place_medal: 

Find below some comments for further improvement. We should merge this PR sooner than later to prevent merge conflicts.

logmanoriginal commented:
Regarding the enclosures: Can you open a separate issue for that? 

I think the current implementation, while not perfect for your needs, does work as intended. Further changes should be separated from this PR to keep it manageable.

ORelio committed:
[Pull] Implementing changes for #802

 - Fix coding style issues with str append
 - Remove useless CACHE_TIMEOUT
 - Use count() instead of $limit
 - Use defaultLinkTo() + handle strings
 - Use http_build_query()
 - Fix missing </em>
 - Remove error_reporting(0)
 - warning CSS (@LogMANOriginal)
 - Fix typo in FeedExpander comment

ORelio commented:
OK so an issue has been opened regarding thumbnails, and requested changes have been implemented.

Please check that I did not forget anything :wink:

ORelio committed:
[Main] More documentation for markdownToHtml

See #802 for more details

logmanoriginal commented:
As far as I can tell, everything is done :1st_place_medal: 

@teromene If you are alright with the changes, this should get merged.

teromene commented:
For thumbnails, there is support in MRSS : http://www.rssboard.org/media-rss#media-thumbnails
What we can do, is create a new array for thumbnails, that we would merge with the enclosures for standard ATOM

teromene commented on bridges/AnidexBridge.php:
```diff
@@ -0,0 +1,207 @@
+
+		// Retrieve torrent listing from search results, which does not contain torrent description
+		$html = getSimpleHTMLDOM($search_url, array(), $opt)
+		or returnServerError('Could not request Anidex: '.$search_url);
```
```PHP
or returnServerError('Could not request Anidex: ' . $search_url);
```

ORelio replied:
Fixing :+1:

teromene commented on bridges/AnidexBridge.php:
```diff
@@ -0,0 +1,207 @@
+		or returnServerError('Could not request Anidex: '.$search_url);
+		$links = $html->find('a');
+		$results = array();
+		foreach ($links as $link)
```
Why not merge the two loops ?
Moreover, returning an error in case of no element being found is probably not the right way to go: If I am searching for something that I know will appear in the future but does not yet (like a new release), the bridge will fail

ORelio replied:
Seems legit, I wrote this with the idea of failing on purpose if HTML changes in result page, because no results e.g. for a category is not supposed to happen. I'll change it to:
```php
if (empty($results) && empty($this->getInput('q')))
	returnServerError('No results from Anidex: '.$search_url);
```
This way, it will not fail when a keyword is specified.

The loops are not merged because I think it makes the bridge easier to read and maintain. First it eliminates duplicates, then it checks for empty results, and finally it process the results. Is it OK to keep it this way?

teromene replied:
Yeah, I was just wondering if there was a specific reason

teromene commented on bridges/CNETBridge.php:
```diff
@@ -0,0 +1,109 @@
+			}
+
+			$article_title = trim($element->find('h2, h3', 0)->plaintext);
+			$article_uri = self::URI . (substr($element->find('a', 0)->href, 1));
```
No need for the ```(```

ORelio replied:
Indeed. Fixing :+1:

teromene commented on bridges/GBAtempBridge.php:
```diff
@@ -76,70 +84,56 @@ public function collectData(){
 				$content = $this->fetchPostContent($url, self::URI);
-				$this->items[] = $this->buildItem($url, $title, $author, $time, $content);
+				$this->items[] = $this->buildItem($url, $title, $author, $time, $img, $content);
+				unset($newsItem);
```
This should not be necessary

ORelio replied:
When testing the bridge, I had memory exhaustion issues using default values from my PHP instance, and manually unsetting the item helped resolving that because I think otherwise the item stays in memory for a bit before garbage collection occurs. I'll add a comment in source code about this problem.

teromene commented on bridges/GBAtempBridge.php:
```diff
@@ -76,70 +84,56 @@ public function collectData(){
 				$content = $this->cleanupPostContent($intro . $review . $subheader . $procons . $scores, self::URI);
-				$this->items[] = $this->buildItem($url, $title, $author, $time, $content);
+				$this->items[] = $this->buildItem($url, $title, $author, $time, $img, $content);
+				unset($reviewItem);
```
Should not be necessary as well

teromene commented on bridges/GBAtempBridge.php:
```diff
@@ -76,70 +84,56 @@ public function collectData(){
 				$content = $this->fetchPostContent($url, self::URI);
-				$this->items[] = $this->buildItem($url, $title, $author, $time, $content);
+				$this->items[] = $this->buildItem($url, $title, $author, $time, null, $content);
+				unset($tutorialItem);
```
Same

teromene commented on bridges/GBAtempBridge.php:
```diff
@@ -76,70 +84,56 @@ public function collectData(){
 				$content = $this->fetchPostContent($url, self::URI);
-				$this->items[] = $this->buildItem($url, $title, $author, $time, $content);
+				$this->items[] = $this->buildItem($url, $title, $author, $time, null, $content);
+				unset($postItem);
```
Same

teromene commented on bridges/NyaaTorrentsBridge.php:
```diff
@@ -0,0 +1,127 @@
+			if (strpos($link->href, '/view/') === 0 && !in_array($link->href, $results))
+				$results[] = $link->href;
+		if (empty($results))
+			returnServerError('No results from Nyaa: ' . $url, 500);
```
Same as the previous bridge, in case of non-matching search, you should just return an empty feed

ORelio replied:
Fixing the same way as the other bridge :+1:

teromene commented on bridges/NyaaTorrentsBridge.php:
```diff
@@ -0,0 +1,127 @@
+			if ($torrent_id != 0 && ctype_digit($torrent_id)) {
+
+				//Retrieve data for this torrent ID
+				$item_uri = self::URI . 'view/'.$torrent_id;
```
```PHP
$item_uri = self::URI . 'view/' . $torrent_id;
```

ORelio replied:
Fixing :+1:

teromene commented on bridges/YGGTorrentBridge.php:
```diff
@@ -111,7 +111,7 @@ public function collectData() {
 			$count++;
 			if($count == 1) continue;
-			if($count == 12) break;
+			if($count == 22) break;
```
Maybe name the constant or add a comment stating why this particular value is used

ORelio replied:
I did not write this bridge but I guess 1 is for skipping table header, and 12 is because increment occurs before checking the count and we skip 1 item... so in order to process 10 items, we need to stop counting at 12. I increased the count to 22 in order to process 20 items because torrent are frequently uploaded on YGGTorrent and 10 wasn't enough to keep up. Adding comments.

teromene replied:
I am to blame actually... :cry: 
I though there was a specific reason, but indeed this is just for the limit of items.

teromene commented on bridges/ZDNetBridge.php:
```diff
@@ -160,143 +160,44 @@ class ZDNetBridge extends BridgeAbstract {
-				. $contents;
+		$article = getSimpleHTMLDOMCached($item['uri']);
+		if(!$article) {
+			$item['content'] .= '<p><em>Could not request ZDNet: ' . $url . '</em></p>';
```
Here, you should return an error except if there is a valid reason why one specific element is not fetched. In this case, just don't return the article

ORelio replied:
That was to keep consistency with the failsafe behavior explained [here](https://github.com/RSS-Bridge/rss-bridge/pull/802#discussion_r216135762), in order to keep the bridge working in reduced functionality mode when only item contents fails to be fetched. Is it bad?

teromene replied:
I am not really sure about this, even with the previous behaviour...
While it makes sense to not fail when we try to fetch an element, I think returnServerError makes more sense syntactically...
I still stand by my original opinion, however  I understand your point. What do you think @LogMANOriginal ?

ORelio replied:
As this is a broader discussion about how to implement bridges (not just this one), I can stick back to `returnServerError` for now and maybe we can discuss this at a later time using a separate issue?

teromene replied:
Sure.
I'll open it, I wouldn't want to hold this PR any longer.

ORelio replied:
Changed to `returnServerError`.

teromene reviewed (changes requested):
After all this is fixed, I am ok with merging it !

ORelio committed:
[Pull] Additional fixes for #802

teromene commented:
Thanks again for the tremendous work that went into this PR !

ORelio commented:
Thanks also for the detailed review work, I did not expect to have so much feedback! Now that I'm up to date with the main codebase, further contributions should be much smoother. :slightly_smiling_face:
